### PR TITLE
Speed up AOCS index build by only scan needed columns.

### PR DESCRIPTION
**Pull request by wangxuqi**
## Speed up AOCS index build by only scan needed columns.
On first index creation (block directory is created) and only that time all columns are scanned. Later any index creation just scans needed columns.

### CODE REVIEW

**Comment by wangxuqi** [comment]
## Beach Mark
I use 1GB tpch lineitem (6 001 215 rows data) to test index build speed up.
Before this pr, index build on lineitem (l_orderkey) needs almost 4.7s 
```
create index ON lineitem (l_orderkey);
CREATE INDEX
Time: 4677.420 ms (00:04.677)
```
After this ps, index build on lineitem (l_orderkey) needs almost 1.7s
```
create index ON lineitem (l_orderkey);
CREATE INDEX
Time: 1746.630 ms (00:01.747)
```

The wider the table is, the better the performance is.

File: src/backend/access/aocs/aocsam_handler.c [hunk 1]
```diff
@@ -1352,12 +1369,50 @@ aoco_index_build_range_scan(Relation heapRelation,
...
+										 snapshot,		/* snapshot */
+										 0,		/* number of keys */
+										 NULL,		/* scan key */
+										 true,			/* buffer access strategy OK */
+										 allow_sync);	/* syncscan OK? */
+		}
+		else
+		{
+			/*
+			 * if block directory has created, we can only scan needed column.
```
Comment (interma) [inline src/backend/access/aocs/aocsam_handler.c:1387]: just in case, please test these corner cases:
* create index in a txn, and rollback (not sure whether blockdir exists)
* create index and delete it (blockdir should exist)

thanks!

Comment (wangxuqi) [reply src/backend/access/aocs/aocsam_handler.c:1387]: OK. I will add some more test cases.

File: src/include/access/tableam.h [hunk 2]
```diff
@@ -815,6 +815,26 @@ table_beginscan_strat(Relation rel, Snapshot snapshot,
 	return rel->rd_tableam->scan_begin(rel, snapshot, nkeys, key, NULL, flags);
 }
 
+/*
+ * Like table_beginscan_strat(), but table_beginscan_strat_extractcolumns()
+ * extract columns for scan from target list and qualifications.
+ * This is mainly for AOCS tables.
+ */
+static inline TableScanDesc
+table_beginscan_strat_extractcolumns(Relation rel, Snapshot snapshot,
```
Comment (interma) [inline src/include/access/tableam.h:824]: I think we should reuse the existed function: `table_beginscan_es`.
Please take a look.

Comment (wangxuqi) [reply src/include/access/tableam.h:824]: It seems there is no different wether SO_ALLOW_SYNC is true or false during AOCS index build range scan. I'm not sure. If you are sure that is OK, I will change `table_beginscan_strat_extractcolumns` to `table_beginscan_es`

**Review by interma — APPROVED** [review]
LGTM.
Thanks Xuqi!

**Comment by ashwinstar** [comment]
@interma is this PR ready and in good shape to be merged?

**Comment by interma** [comment]
yes, merge it now.
